core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+# than threshold times then vocabulary ignores the token.
+VOCABULARY_THRESHOLD = 5
+
+
+def get_tokens(program):
+    """Generate tokens for program using tokenize module.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+
+    Yields:
+        int. Integer value which represents type of token.
+        str. Token.
+    """
+    for token_id, token_name, _, _, _ in tokenize.generate_tokens(
+            StringIO.StringIO(program).readline):
+        yield (token_id, token_name)
+
+
+def cv_tokenizer(program):
+    """Custom tokenizer for tokenizing Python programs.
+
+    This needs to be passed as an argument to sci-kit's CountVectorizer.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+        VAR_TOKEN: str. Replace variable and method names with this string when
+            tokenizing data.
+
+    Returns:
+        list(str). A list containing tokens of input program.
+    """
+    tokenized_program = []
+    for token_id, token_name in get_tokens(program):
+        # Ignore all newline, comment and empty string tokens.
+        # token_id is token.N_TOKENS for empty line / newline token.
+        # token_id is 54 for comments.
+        if (token_id == token.N_TOKENS or token_id == COMMENT_TOKEN
+                or token_name.strip() == ''):
+            continue
+        elif token_id == token.NAME:
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword. Treat all variables and methods similar.
+            if token_name in keyword.kwlist:
+                tokenized_program.append(token_name)
+            else:
+                tokenized_program.append(VAR_TOKEN)
+        else:
+            tokenized_program.append(token_name)
+
+    return tokenized_program
+
+
+def map_tokens_to_ids(training_data, threshold):
+    """Generates a list of valid tokens and assigns a unique ID to each
+    token.
+
+    Args:
+        training_data: dict. A dictionary containing training_data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program
+                    } for each program in training_data.
+                }
+        threshold: int. Ignore a token if it appears less than 'threshold' times
+            in dataset.
+
+    Returns:
+        dict. A dictionary containing tokens mapped to a unique ID.
+    """
+    # All unique tokens and number of time they occur in dataset. A token
+    # can be a keyword, an identifier, a constant, an operator.
+    vocabulary = {}
+
+    for pid in training_data:
+        program = training_data[pid]['source']
+        for token_id, token_name in get_tokens(program):
+            # Ignore all newline, comment and empty string tokens.
+            if (token_id == token.N_TOKENS or token_id == COMMENT_TOKEN
```

Comment:
We seem to use this construct a lot. It may be worth pulling out into a separate function so that you can do `if token_is_ignorable(token_id, token_name)` or similar.